### DISCUSSION TITLE
Encrypting the database in the web implementation

### DISCUSSION THREAD

**Discussion by majdi21** [simolus3/drift#1694] [Q&A]
Hello,

   Did anybody try to use an encrypted database in the web? With Flutter web in full force, this would be an amazing addition as it will allow for a full PWA with local storage that is encrypted and safe. Please share anything you have tried for this (whether the attempt was successful or not).

Thank you

**Comment by simolus3** [comment]
You can compile a custom version of sql.js with support for encryption, such as described [here](https://github.com/sql-js/sql.js/issues/54#issuecomment-480712412). By replacing sql.js with the one with support for encryption and by running the encryption pragmas when opening the database, you can encrypt the database on the web.

**Reply by jeannelim** [reply]
Hi @simolus3, how do I run the encryption pragmas when opening the database on web? The WebDatabase constructor doesn't have the setup parameter like in NativeDatabase.

**Reply by simolus3** [reply]
Good point. We don't currently expose this functionality (and it would be a bit hard to do so since we don't have a public Dart API on the raw sql.js interface).

You could technically put it in `onCreate` as well, just before you start using the database?

```dart
MigrationStrategy get migration {
  return MigrationStrategy(
    onCreate: (m) async {
      await customStatement('pragma key = ...');
      await m.createAll();
    },
  );
}
```

**Reply by jeannelim** [reply]
Thanks, I tried this with the compiled custom version of sql.js (replaced sqlite with sqleet). Now when I run it on the web and use local storage, I can see that the database is encrypted but I also get this error:

Error: file is not a database

    Here is the stack trace:
    ========================
    sql-wasm.js 90:155                                                   handleError
sql-wasm.js 88:420                                                   prepare
dart-sdk/lib/_internal/js_dev_runtime/patch/js_patch.dart 143:53     callMethod
packages/drift/src/web/sql_js.dart.js 89:44                          prepare
packages/drift/web.dart.js 515:34                                    runSelect
packages/drift/src/utils/lazy_database.dart.js 7559:28               <fn>
packages/drift/src/utils/lazy_database.dart.js 7536:18               <fn>
dart-sdk/lib/async/future.dart 296:31                                sync
packages/drift/src/utils/synchronized.dart.js 28:33                  callBlockAndComplete
packages/drift/src/utils/synchronized.dart.js 32:46                  <fn>
dart-sdk/lib/async/zone.dart 1434:47                                 _rootRunUnary
dart-sdk/lib/async/zone.dart 1335:19                                 runUnary
dart-sdk/lib/async/future_impl.dart 159:18                           handleValue
dart-sdk/lib/async/future_impl.dart 766:44                           handleValueCallback
dart-sdk/lib/async/future_impl.dart 795:13                           _propagateToListeners
dart-sdk/lib/async/future_impl.dart 601:5                            [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 639:7                            <fn>
dart-sdk/lib/async/zone.dart 1426:13                                 _rootRun
dart-sdk/lib/async/zone.dart 1328:19                                 run
dart-sdk/lib/async/zone.dart 1236:7                                  runGuarded
dart-sdk/lib/async/zone.dart 1276:23                                 callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                     _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                      _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 166:15  <fn>

**Reply by jeannelim** [reply]
Hi @simolus3, I created a sample app with the wasm files I generated: https://github.com/jeannelim/drift_sample
Even in the sample app I get the error "file is not a database". In my app, I managed to remove some of it by calling the pragma key before an Insert or Select call to the database like this: db.customStatement('PRAGMA key = password');
I tried the same in the sample app but I still get the error. Do you have some ideas on how to get rid of the error?

**Reply by simolus3** [reply]
In the next drift version, we'll expose the wrapper around the sql.js interface and provide a proper `setup` parameter for web databases as well. I think this should solve the problem, I'll try out your sample with the development version of drift and see if that can fix the problem. Thanks for setting that up!

In the longer term, I'm also working on a version of the `sqlite3` package for the web in https://github.com/simolus3/sqlite3.dart/pull/80. That one will support the exact same API as the `NativeDatabase` does now, without requiring any JavaScript or emscripten glue code.

**Reply by simolus3** [reply]
Just took a look at your example. I can confirm it works with drift from 35efc68e42214341d5bdf99a26f4f0a9db5a8bfe if I remove the key pragma from the migration and instead do

```dart
final db = Database(WebDatabase('sample', logStatements: true, setup: (db) {
  db.run("pragma KEY = 'password';");
}));
```

Hopefully this situation will get a lot cleaner with the upcoming wasm backend for the `sqlite3` package :D

**Reply by jeannelim** [reply]
That's great! When is the next release?

**Reply by jeannelim** [reply]
I tried commit 35efc68 and I still get the error but only on the first select/insert statement after the database is created. So I tried calling doWhenOpened right after the database is created eg. await db.doWhenOpened((e) => print('db ready'));
That got rid of the error and it hasn't popped up since.

**Reply by jeannelim** [reply]
Hi @simolus3. I saw this note in the WebDatabase constructor "With the current web implementation, the database is closed and re-opened whenever it is stored (i.e. after every transaction or insert/update/delete outside of transactions). The [setup] parameter will also be invoked whenever the database is re-opened." This gave me the idea to use a transaction to improve performance. Then I got the error File is not a database again after the transaction executes and another query runs. It happens when I use indexedDb for storage, local storage seems fine.

**Reply by jeannelim** [reply]
Hi @simolus3, you mentioned earlier that you are working on a version of the sqlite3 package for the web. Do you have a timeline on the new package? Thanks.

**Reply by simolus3** [reply]
> Do you have a timeline on the new package? Thanks.

The current status is that the same "File is not a database" error occurs with that new version too. I don't fully understand how that can happen at the moment. I'll need to carefully compare the persistence logic on the web with sqlite3's builtin file implementations, and unfortunately I didn't have time for that yet.

**Reply by jeannelim** [reply]
No worries. What I meant to ask was if you have a target completion date for the new implementation? We're trying to decide if we should wait for the new implementation.

**Reply by simolus3** [reply]
I plan to release the new drift version with the `setup` parameter this weekend. The new version of the `sqlite3` package will definitely take longer, I don't have an ETA for that yet. So I would pick what works best at the moment :)
